File: Android/app/src/main/java/knk/ee/neverland/fakeapi/FakeAuthAPI.java
package knk.ee.neverland.fakeapi;

import com.google.common.hash.Hashing;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

import knk.ee.neverland.api.AuthAPI;
import knk.ee.neverland.pojos.RegistrationData;
import knk.ee.neverland.exceptions.LoginFailedException;
import knk.ee.neverland.exceptions.RegistrationFailedException;

public class FakeAuthAPI implements AuthAPI {
    private List<User> registeredUsers = new ArrayList<>();
    private List<String> workingKeys = new ArrayList<>();

    public FakeAuthAPI() {
        workingKeys.add("ABC");
    }

    @Override
    public String attemptLogin(final String login, final String password) throws LoginFailedException {
        for (User user : registeredUsers) {
            if (user.login.equals(login) && user.password.equals(password)) {
                String key = generateKey(user);
                workingKeys.add(key);
                return key;
            }
        }

        throw new LoginFailedException("User not found!");
    }

    @Override
    public void registerAccount(final RegistrationData registrationData) throws RegistrationFailedException {
        if (!loginIsCorrect(registrationData.getLogin())) {
            throw new RegistrationFailedException("Incorrect login.");
        }

        if (!passwordIsCorrect(registrationData.getPassword())) {
            throw new RegistrationFailedException("Incorrect password.");
        }

        if (!nameIsCorrect(registrationData.getFirstName())) {
            throw new RegistrationFailedException("Incorrect first name.");
        }

        if (!nameIsCorrect(registrationData.getSecondName())) {
            throw new RegistrationFailedException("Incorrect second name.");
        }

        if (!emailIsCorrect(registrationData.getEmail())) {
            throw new RegistrationFailedException("Incorrect email.");
        }

        User user = new User();

        user.login = registrationData.getLogin();
        user.password = registrationData.getPassword();
        user.firstName = registrationData.getFirstName();
        user.secondName = registrationData.getSecondName();
        user.email = registrationData.getEmail();

        registeredUsers.add(user);
    }

    @Override
    public boolean isKeyActive(String key) {
        return workingKeys.contains(key);
    }

    private String generateKey(User user) {
        String line = String.format("[%s==%s]", user.login, user.password);
        return Hashing.sha256()
                .hashString(line, StandardCharsets.UTF_8)
                .toString();
    }

    private boolean loginIsCorrect(String login) {
        return login.matches("^[a-z0-9_-]{6,16}$");
    }

    private boolean passwordIsCorrect(String password) {
        return password.matches("^[a-z0-9_-]{6,18}$");
    }

    private boolean emailIsCorrect(String email) {
        return email.matches("^([a-z0-9_\\.-]+)@([\\da-z\\.-]+)\\.([a-z\\.]{2,6})$");
    }

    private boolean nameIsCorrect(String name) {
        return name.matches("^[A-Za-z ,.'-]+$");
    }

    private class User {
        String login;
        String password;
        String email;
        String firstName;
        String secondName;
    }
}

File: Android/app/src/main/java/knk/ee/neverland/fakeapi/FakeAuthAPI.kt
package knk.ee.neverland.fakeapi

import com.google.common.hash.Hashing
import knk.ee.neverland.api.AuthAPI
import knk.ee.neverland.api.AuthAPIResponse
import knk.ee.neverland.exceptions.AuthAPIException
import knk.ee.neverland.pojos.RegistrationData
import java.nio.charset.StandardCharsets
import java.util.*

class FakeAuthAPI : AuthAPI {
    private val registeredUsers = ArrayList<User>()
    private val workingKeys = ArrayList<String>()

    init {
        workingKeys.add("ABC")
    }

    @Throws(AuthAPIException::class)
    override fun attemptLogin(login: String, password: String): String {
        for (user in registeredUsers) {
            if (user.login == login && user.password == password) {
                val key = generateKey(user)
                workingKeys.add(key)
                return key
            }
        }

        throw AuthAPIException(AuthAPIResponse.REQUEST_FAILED)
    }

    @Throws(AuthAPIException::class)
    override fun registerAccount(registrationData: RegistrationData) {
        if (!loginIsCorrect(registrationData.login)) {
            throw AuthAPIException(AuthAPIResponse.REQUEST_FAILED)
        }

        if (!passwordIsCorrect(registrationData.password)) {
            throw AuthAPIException(AuthAPIResponse.REQUEST_FAILED)
        }

        if (!nameIsCorrect(registrationData.firstName)) {
            throw AuthAPIException(AuthAPIResponse.REQUEST_FAILED)
        }

        if (!nameIsCorrect(registrationData.secondName)) {
            throw AuthAPIException(AuthAPIResponse.REQUEST_FAILED)
        }

        if (!emailIsCorrect(registrationData.email)) {
            throw AuthAPIException(AuthAPIResponse.REQUEST_FAILED)
        }

        val user = User()

        user.login = registrationData.login
        user.password = registrationData.password
        user.firstName = registrationData.firstName
        user.secondName = registrationData.secondName
        user.email = registrationData.email

        registeredUsers.add(user)
    }

    override fun isKeyActive(key: String): Boolean {
        return workingKeys.contains(key)
    }

    private fun generateKey(user: User): String {
        val line = String.format("[%s==%s]", user.login, user.password)
        return Hashing.sha256()
                .hashString(line, StandardCharsets.UTF_8)
                .toString()
    }

    private fun loginIsCorrect(login: String): Boolean {
        return login.matches("^[a-z0-9_-]{6,16}$".toRegex())
    }

    private fun passwordIsCorrect(password: String): Boolean {
        return password.matches("^[a-z0-9_-]{6,18}$".toRegex())
    }

    private fun emailIsCorrect(email: String): Boolean {
        return email.matches("^([a-z0-9_\\.-]+)@([\\da-z\\.-]+)\\.([a-z\\.]{2,6})$".toRegex())
    }

    private fun nameIsCorrect(name: String): Boolean {
        return name.matches("^[A-Za-z ,.'-]+$".toRegex())
    }

    private inner class User {
        internal var login: String? = null
        internal var password: String? = null
        internal var email: String? = null
        internal var firstName: String? = null
        internal var secondName: String? = null
    }
}
